src/errors.ts

```diff
@@ -381,11 +395,22 @@ export class ParseError extends SyntaxError {
  * @param {...string[]} params
  * @returns {never}
  */
-export function report(parser: ParserState, type: Errors, ...params: string[]): never {
-  throw new ParseError(parser.index, parser.line, parser.column, type, ...params);
+export function report(parser: ParserState, context: Context, node: Node, type: Errors, ...params: string[]): never {
+  const end = finishNode(parser, context, parser.index, parser.line, parser.column, node).loc?.end;
```

Comment:
The finishNode mutates nothing on `parser: ParserState`. If you look into the implementation, the `loc.end` is just
```
end: {
        line: parser.startLine,
        column: parser.startColumn
      }
```
Which you can directly get from `parser`.
Also, the availability of `loc.end` in finishNode depends on `OptionsLoc`, but you can always get `parser.startLine` directly.